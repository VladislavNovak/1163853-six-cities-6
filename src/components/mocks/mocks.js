export const rawHotel = {
  "city": {
    "name": `Paris`,
    "location": {
      "latitude": 48.85661,
      "longitude": 2.351499,
      "zoom": 13
    }
  },
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
  "images": [`https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`],
  "title": `The house among olive`,
  "is_favorite": false,
  "is_premium": true,
  "rating": 3,
  "type": `room`,
  "bedrooms": 1,
  "max_adults": 1,
  "price": 169,
  "goods": [`Breakfast`],
  "host": {
    "id": 25,
    "name": `Angelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-angelina.jpg`
  },
  "description": `Relax, rejuvenate`,
  "location": {
    "latitude": 48.83861,
    "longitude": 2.350499,
    "zoom": 16
  },
  "id": 1
};

export const rawComment = ({
  "id": 1,
  "user": {
    "id": 11,
    "is_pro": false,
    "name": `Jack`,
    "avatar_url": `https://assets.htmlacademy.ru`
  },
  "rating": 3,
  "comment": `Beautiful space, fantastic location and`,
  "date": `2021-02-10T08:04:28.647Z`
});

export const rawUserInfo = ({
  userEmail: `email`,
  userAvatar: `avatar_url`,
});

export const rawHotels = [rawHotel];
export const rawComments = [rawComment];
