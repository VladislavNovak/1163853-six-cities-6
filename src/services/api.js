import axios from "axios";
import {HttpCode} from "../utils/constants";

const BACKEND_URL = `https://6.react.pages.academy/six-cities`;
const REQUEST_TIMEOUT = 5000;

export const createAPI = (onUnauthorized) => {
  const api = axios.create({
    baseURL: BACKEND_URL,
    timeout: REQUEST_TIMEOUT,
    withCredentials: true,
  });

  const onSuccess = (response) => response;

  const onFail = (err) => {
    const {response} = err;

    if (response.status === HttpCode.UNAUTHORIZED) {
      onUnauthorized();

      // Бросаем ошибку, потому что нам важно прервать цепочку промисов после запроса авторизации.
      // Запрос авторизации — это особый случай и важно дать понять приложению, что запрос был неудачным.
      throw err;
    }
  };

  api.interceptors.response.use(onSuccess, onFail);

  return api;
};

// createAPI - возвращает настроенный объект axios
// withCredentials - указывает, нужно ли использовать куки. В случае этого проекта,
//    авторизация построена на куках, поэтому ключ true
// api.interceptors - перехватывают коллбэки и перенаправляют выполнение на onSuccess или на onFail. Объяснение:
//    в случае с fetch, при возникновении, к примеру, 401, выводится ошибка,
//    axios же позволяет обработать эту ошибку. Т.о. в случае успеха, вызывается onSuccess,
//    извлекается response и перекидывается далее на выполнение. В случае onFail, можно самостоятельно
//    выбрать стратегию действий. В данном проекте мы вызываем коллбэк onUnauthorized, который передаётся
//    в конструктор createAPI впоследстции данный коллбэк будет редиректить пользователя на страницу авторизации
