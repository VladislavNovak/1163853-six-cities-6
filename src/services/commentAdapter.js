const adaptOneCommentToClient = (comment) => ({
  id: String(comment.id),
  quote: comment.comment,
  date: comment.date,
  rating: String(comment.rating),
  visitorId: String(comment.user.id),
  visitorName: comment.user.name,
  visitorAvatar: comment.user.avatar_url,
  visitorIsPro: comment.user.is_pro,
});

export const adaptAllCommentsToClient = (data) => data.map(adaptOneCommentToClient);
