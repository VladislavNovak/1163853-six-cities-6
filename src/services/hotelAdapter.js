export const adaptOneHotelToClient = (hotel) => ({
  id: String(hotel.id),
  cityName: hotel.city.name,
  cityLatitude: hotel.city.location.latitude,
  cityLongitude: hotel.city.location.longitude,
  cityZoom: hotel.city.location.zoom,
  title: hotel.title,
  images: hotel.images,
  preview: hotel.preview_image,
  isPremium: hotel.is_premium,
  isFavorite: hotel.is_favorite,
  bedrooms: hotel.bedrooms,
  adults: hotel.max_adults,
  price: hotel.price,
  rating: hotel.rating,
  description: hotel.description,
  type: hotel.type,
  hostName: hotel.host.name,
  hostAvatar: hotel.host.avatar_url,
  hostId: String(hotel.host.id),
  services: hotel.goods,
  hostIsPro: hotel.host.is_pro,
  latitude: hotel.location.latitude,
  longitude: hotel.location.longitude,
  zoom: hotel.location.zoom,
});

export const adaptAllHotelsToClient = (data) => data.map(adaptOneHotelToClient);
