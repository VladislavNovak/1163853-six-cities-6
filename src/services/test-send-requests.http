# GET https://6.react.pages.academy/six-cities/hotels HTTP/1.1
GET https://6.react.pages.academy/six-cities/comments/4 HTTP/1.1
# Authorization: Basic NS43cqsibarQ5EJ
Content-Type: application/json
