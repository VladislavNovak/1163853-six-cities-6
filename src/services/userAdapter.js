export const userAdapter = (data) => ({
  userEmail: data[`email`],
  userAvatar: data[`avatar_url`],
});
