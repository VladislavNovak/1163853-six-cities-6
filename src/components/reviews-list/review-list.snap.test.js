import React from 'react';
import {render} from '@testing-library/react';
import {rawComments} from '../mocks/mocks';
import {adaptAllCommentsToClient} from '../../services/commentAdapter';

import {ReviewsList} from '..';

test(`Should ReviewsList render correctly`, () => {
  const comments = adaptAllCommentsToClient(rawComments);

  const {container} = render(<ReviewsList comments={comments}/>);
  expect(container).toMatchSnapshot();
});
