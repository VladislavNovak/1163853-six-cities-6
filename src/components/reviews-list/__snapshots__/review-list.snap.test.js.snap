// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ReviewsList render correctly 1`] = `
<div>
  <h2
    class="reviews__title"
  >
    Reviews Â· 
    <span
      class="reviews__amount"
    >
      1
    </span>
  </h2>
  <ul
    class="reviews__list"
  >
    <li
      class="reviews__item"
    >
      <div
        class="reviews__user user"
      >
        <div
          class="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            class="reviews__avatar user__avatar"
            height="54"
            src="https://assets.htmlacademy.ru"
            width="54"
          />
        </div>
        <span
          class="reviews__user-name"
        >
          Jack
        </span>
      </div>
      <div
        class="reviews__info"
      >
        <div
          class="reviews__rating rating"
        >
          <div
            class="reviews__stars rating__stars"
          >
            <span
              style="width: 60%;"
            />
            <span
              class="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          class="reviews__text"
        >
          Beautiful space, fantastic location and
        </p>
        <time
          class="reviews__time"
          datetime="2021-02-10"
        >
          February 2021
        </time>
      </div>
    </li>
  </ul>
</div>
`;
